### PR TITLE
feat: add contains anyof and contains allof operators

### PR DESCRIPTION
**Pull request by arththebird**
## Description

Add `contains anyof` and `contains allof` operators.

## Checklist

- [X] PR title follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) guidelines 
- [X] All relevant PR sections are populated, irrelevant ones are removed 

## Validation
- [X] All tests pass

image

### CODE REVIEW

File: pypred/ast.py [hunk 1]
```diff
@@ -307,38 +307,96 @@ def failure_info(self, ctx):
...
+        self.type = op
         self.left = left
         self.right = right
 
+    def name(self):
+        return "%s operator at %s" % (self.type.upper(), self.position)
+
     def _validate(self, info):
-        if not isinstance(self.right, (Number, Literal, Constant)):
+        if self.type not in ("contains", "any", "all"):
```
Comment (edouardlp) [inline pypred/ast.py:321]: Does pypred have any constants/enums to declare types? It seems odd/error prone to use the string everywhere.

File: pypred/parser.py [hunk 2]
```diff
@@ -12,6 +12,8 @@
     'and': 'AND',
     'or': 'OR',
     'not': 'NOT',
+    'anyof': 'ANY',
```
Comment (edouardlp) [inline pypred/parser.py:15]: Have you found any supported operators in pypred that are in 2 words? I'm having a hard time understanding how `contains` and `anyof`/`allof` can be registered as distinct operators, and have all of this working. Doesn't the AST become something like : 
`left contains anyof right`

```
           contains
left -                   - anyof
                                             - right
```

This seems a bit odd.
Isn't there a way to register `contains anyof` as a single operator such that.

```
           contains anyof
left -                   - right
```
?

File: pypred/ast.py [hunk 3]
```diff
@@ -307,38 +307,96 @@ def failure_info(self, ctx):
...
-        if hasattr(left, "__contains__"):
-            right = self.right.eval(ctx)
+        if not hasattr(left, "__contains__"):
+            return False
+
+        right = self.right.eval(ctx)
+        if self.type == "contains":
             return right in left
+
+        if not hasattr(right, "__iter__"):
```
Comment (edouardlp) [inline pypred/ast.py:356]: Why is this not needed for `left`? What makes this check for left?

File: tests/integ/preds.txt [hunk 4]
```diff
@@ -25,3 +25,7 @@ thingy is 200 and val is 100 # false
 {"Jack" "Joe" "Jill"} contains name # true
 {400 500} contains status # true
 {"east-db-001" "east-foo-002"} contains server # false
+errors contains anyof {'cpu load' 'foo'} # true
+errors contains anyof {'foo' 'bar'} # false
```
Comment (edouardlp) [inline tests/integ/preds.txt:29]: Have you thought about what happens with `error not contains anyof` 🤯 ?

File: pypred/compact.py [hunk 5]
```diff
@@ -32,10 +32,10 @@ def node_name(node):
     elif cls_name == "NegateOperator":
         return (cls_name, node_name(node.left))
 
-    elif cls_name in ("CompareOperator", "LogicalOperator"):
+    elif cls_name in ("CompareOperator", "LogicalOperator", "ContainsOperator"):
```
Comment (edouardlp) [inline pypred/compact.py:35]: IIUC, this is the approach pypred uses for polymorphism? I.e `CompareOperator` is the class, and the type is provided by `node.type` so that the class can know to compare greater than equal, less than, etc.? Hence we do this because we need to know the type of contains (regular vs contains any vs contains anyof)

File: pypred/ast.py (hunk 1 above)
Comment (arththebird) [reply pypred/ast.py:321]: From what I've seen, no. I agree it's pretty bad and error prone, but I tried to stay in line with what was there. If we don't want to merge back to the original project, I have a lot of improvements in mind that can be done 😁 

Without going into the "refactor everything" route, I could introduce a constants file for these new ones

Comment (edouardlp) [reply pypred/ast.py:321]: Haha fair, no this is fine. I agree to keep the `pypred` style as much as possible unless we don't get any feedback from the maintainers.

File: pypred/parser.py (hunk 2 above)
Comment (arththebird) [reply pypred/parser.py:15]: There is `is not` that's in two words, and that's the only one I think. 

This is for registering a reserved word in the parser so that it can be mapped as a specific token and not a `STRING` token (see `t_STRING(t)` function in parser.py) [Link](https://github.com/dialoguemd/pypred/pull/2/files#diff-8a01c43e06ad9126fa1060076a598a9abde2ec91bbbed0a75a7ee1fff0882295R84).

The magic happens in the functions starting with `p_`. In the `p_contains_any(p)` function, the documentation line (`term : factor CONTAINS ANY factor` ) tells the parser that `factor CONTAINS ANY factor` should be mapped to a term and when meeting that term, execute the body of the function. This is the way `ply` is working, so I don't see another way to do it without deep diving in `ply` implementation.

File: pypred/ast.py (hunk 3 above)
Comment (arththebird) [reply pypred/ast.py:356]: It's because we don't iterate on `left`, so we just check that `hasattr(left, "__contains__")` since this is the method that gets executed in `x in y` IIUC

File: tests/integ/preds.txt (hunk 4 above)
Comment (arththebird) [reply tests/integ/preds.txt:29]: Good point, will try next week 😅

File: pypred/compact.py (hunk 5 above)
Comment (arththebird) [reply pypred/compact.py:35]: Exactly! I went that way from what I've seen in the project, but another way would be to create a class for each new operators and use them instead of using the `self.type` attribute to determine to operation to be done

File: tests/integ/preds.txt (hunk 4 above)
Comment (arththebird) [reply tests/integ/preds.txt:29]: So it works, but the syntax is `not error contains anyof` 🙂

Comment (edouardlp) [reply tests/integ/preds.txt:29]: Cool. What happens if you invert? (i.e `error not contains anyof`)?

Comment (arththebird) [reply tests/integ/preds.txt:29]: It throws an error saying that the predicate is invalid with an explanation like `Syntax error with not at line 1, col 8`

Comment (edouardlp) [reply tests/integ/preds.txt:29]: Ok, this should be good for our purposes then.

File: pypred/ast.py (hunk 3 above)
Comment (edouardlp) [reply pypred/ast.py:356]: I'm not sure I follow, we do iterate over both `left` and `right` in [this expression](https://github.com/dialoguemd/pypred/pull/2/files#diff-ff6c318c7d52ada0696e22d0127c9cd6e215e90e9415d173bf1f6a4835876faaR359)

Comment (arththebird) [reply pypred/ast.py:356]: Actually, the way I understand it is that when you do `x in y` in python, the `__contains__` method gets executed. 

So for the expression `any(right_item in left for right_item in right)`, we iterate over `right`, for each items we check that `right_item in left` and the `any` function exits and return `True` as soon as there is a `True` returned. So we never call the `__iter__` function on the `left` parameter

File: tests/unit/test_ast.py [hunk 6]
```diff
@@ -93,13 +93,26 @@ def test_match_bad_arg(self):
...
     def test_contains_bad(self):
-        a = ast.ContainsOperator(ast.Literal("foo"), ast.Empty())
+        a = ast.ContainsOperator("contains", ast.Literal("foo"), ast.Empty())
         valid, info = a.validate()
         assert not valid
         assert "Contains operator must take" in info["errors"][0]
 
     def test_contains_valid_args(self):
-        a = ast.ContainsOperator(ast.Literal("foo"), ast.Literal("bar"))
+        a = ast.ContainsOperator("contains", ast.Literal("foo"), ast.Literal("bar"))
```
Comment (edouardlp) [inline tests/unit/test_ast.py:102]: I'm not sure I understand why this is valid. IIUC, this is because `pypred` allows you to use contains to compare strings litterals?

File: pypred/ast.py (hunk 3 above)
Comment (edouardlp) [reply pypred/ast.py:356]: Ah yes my bad, makes sense. `for in` and `in` are obviously different operators.

File: tests/unit/test_ast.py (hunk 6 above)
Comment (arththebird) [reply tests/unit/test_ast.py:102]: This is valid because `Literal` nodes are used for literal string values (`'foo'`) and for string that are used as an identifier and resolve to another value (`some_list_slot`). For example: `some_list_slot contains 'foo'`

Comment (arththebird) [reply tests/unit/test_ast.py:102]: I decided to not handle the case for `contains anyof` and `contains allof` to reduce complexity, so we couldn't do something like `{'foo' 'bar'} contains allof some_list_slot`

Comment (arththebird) [reply tests/unit/test_ast.py:102]: I admit that this is not great, in my opinion nodes that are identifiers should be different than nodes for string literals, but that's how it's handled currently in the library 😕

Comment (edouardlp) [reply tests/unit/test_ast.py:102]: I see. This seems fine temporarily, but it's not intuitive. I expect to be able to use identifiers and literals interchangeably everywhere. I suggest we fix that in the next story. I do agree with you, the best solution for this would be to deal with identifiers in the same way everywhere, having a node type for it makes it much clearer/easier to deal with.

File: tests/unit/test_ast.py [hunk 7]
```diff
@@ -265,49 +278,126 @@ def test_compare_empty(self, type):
...
 
-    def test_contains_invalid(self):
+    @pytest.mark.parametrize("type", ["contains", "any", "all"])
+    def test_contains_invalid(self, type):
         l = ast.Literal("l")
         r = ast.Literal("r")
-        a = ast.ContainsOperator(l, r)
-        d = {"l": 1, "r": None}
+        a = ast.ContainsOperator(type, l, r)
+        d = {"l": 1, "r": 1}
```
Comment (edouardlp) [inline tests/unit/test_ast.py:286]: Why this change?

Comment (arththebird) [reply tests/unit/test_ast.py:286]: Because I felt like the test was misleading. When looking at it, I felt like it was invalid because of the `None` but actually, it's because `l` does not support `__contains__`, so the `None` value doesn't bring anything to the test.

File: tests/unit/test_ast.py (hunk 6 above)
Comment (edouardlp) [reply tests/unit/test_ast.py:102]: Sorry another question to clarify : Can we do `some_set_slot contains anyof some_other_set_slot`?

Comment (edouardlp) [reply tests/unit/test_ast.py:102]: If I understand [this](https://github.com/dialoguemd/pypred/pull/2/files#diff-9f0c6436c1f69ab65cf72a07463aee64c776009f8f9a797d74590b3ebf577c8eR107) correctly, I think no?

File: pypred/parser.py (hunk 2 above)
Comment (edouardlp) [reply pypred/parser.py:15]: As discussed at standup, let's address this in a separate story.

File: tests/unit/test_ast.py (hunk 6 above)
Comment (arththebird) [reply tests/unit/test_ast.py:102]: For the moment no, it's not allowed, but will be in next story while doing 👇  

> I expect to be able to use identifiers and literals interchangeably everywhere. I suggest we fix that in the next story.

**Review by edouardlp — APPROVED** [review]
🍳 👏

As discussed in the comments, this is a first implementation that does not handle all expected cases, mostly towards the use of _identifiers_. It also does not produce an AST that is in an expected shape. As such is not ready to be merged in to `pypred` but is in a great shape to be used in our project to unlock the next steps.

Let's come back to this next sprint!
